### PR TITLE
MNT: Re-rendered with conda-smithy 2.2.1

### PR DESCRIPTION
**Pull request by tritemio**
This adds python 3.6 to the build matrix, and removes python 3.4.

**Note:** Until now, only lmfit 0.8.3 was built for python 3.6 on conda-forge. This caused python 3.6 users to silently install an outdated lmfit version. This PR should fix the issue.

### CODE REVIEW

**Comment by ericdill** [comment]
The windows failure was due to a transient build failure in the emcee build on appveyor, discussed here: https://github.com/conda-forge/conda-forge.github.io/issues/364. Once the emcee merge to master completes, we can restart the Appveyor build here which should be working just fine.

**Comment by tritemio** [comment]
@ericdill, thanks for your help.

I did this PR to build lmfit 0.9.5 for python 3.6 and I didn't merge it due to the failing python 3.6 build. In the meanwhile @newville released lmfit 0.9.6, updated this feedstock and merged to master (with failing python 3.6 build).

What do you think is the best path forward to have both lmfit 0.9.5 and 0.9.6 built for python 3.6.

Thanks.

**Comment by newville** [comment]
@tritemio @ericdill  I think it might be sufficient make a new PR with a trivial merge, no?  That is, it was the upstream package that was causing the odd failure.

**Comment by tritemio** [comment]
Yes, I guess after have the package built for lmfit 0.9.6 from the last commit on master, I can merge a rebased PR to build lmfit 0.9.5. Then master will have the recipe for lmfit 0.9.5 until the next release.

**Comment by newville** [comment]
@tritemio I meant that I think (but could be wrong) that *any* new commit to master here should trigger a fresh rebuild and generate new conda results.  I'll try that....

**Comment by tritemio** [comment]
@newville, True. I'll wait for lmfit 0.9.6 to be built for win64 python 3.6, then open a PR for 0.9.5. Builds on master branch automatically publish the packages. The concept of building through a PR first is to test build process before releasing the conda package.

**Comment by ericdill** [comment]
Pushes to *any* branch on conda-forge/lmfit-feedstock build and push to
anaconda.org/conda-forge. You can push a commit with 0.9.5 to a new branch
while the master branch can remain at 0.9.6. Hopefully this makes sense.
I'm typing this on my phone. I can elaborate as needed.

**Comment by newville** [comment]
To be clear, I have no problem pushing anything to master on conda-forge/lmfit-feedstock.  Do I care about a clean history for the repository for the conda recipe?  No, I do not.  

The latest version of the lmfit recipe (actually identical to build 216df66d32cbb19d5d4e3fa058ec45c725727c94 ) failed with Python 3.6 on TravisCI, with an error from `conda build` when trying to install the required packages.  That error  is completely independent of the lmfit code -- the build never got to the point of downloading it.

The same recipe succeeded on the previous commit a few days ago. And, for sure, lmfit on Mac OSX with Anaconda Python 3.6 passes all tests -- most of the devel done on it over the past month has been on that platform.   

The appveyor tests have been Queued for two hours.  Yep.

I see the appeal of conda-forge, but question it's actual utility.  It's test mechanism is highly inaccurate, with far too many false negatives.  That's sort of absurd for a pure python package such as lmfit.  To be clear, the package is not in question -- it passes all tests on all the platforms in question, including when using  Anaconda.  It is only the conda recipe and its own assumptions and test mechanism that is in question.

I've experienced similar in nature (but completely different in detail) with a conda recipe for another project -- false negatives that are not due to problems with the package itself, but with problems with the assumptions conda-forge makes about building and testing.   

In short, conda-forge sucks.   Whether it is worth the effort to make recipes that work for it is an open question.  Certainly, if Travis is failing before it even downloads lmfit, we're not going to solve that problem here.   

Hurray for `pip install`.

**Comment by newville** [comment]
@ericdill   Sorry if using the word "sucks" sounds to you like open hostility.   I did not and do not mean to imply ill-will to anyone here or in any way associated with the project.   Calling software broken does not in any way imply that the people writing the software have any attributes that would disqualify them for sainthood.  I break things all the time.  I write lots of software, some of it sucks.  I try to fix the problems, and  often try to help others fix broken software. 

This package is pure python, reasonably well tested on many platforms and python versions, and actually depends on only on packages from base anaconda. The conda-forge recipe for lmfit is failing to build solely due to problems with conda-forge itself.  I assume these will be fixed.  And I assume that Appveyor will eventually work to build the missing win-64/Py3.6 package (though it's now been queued for 4 hours!). 

Until then, `pip install` works on all platforms.   And, until it can deal with a package and recipe as simple as lmfit, I think it is completely defensible to say that conda-forge is broken.

**Comment by tritemio** [comment]
@newville:
A conda package for lmfit is useful if you (a) use conda/anaconda, (b) have lmfit as dependency of other conda packages. If lmfit is not a dependency then installing it by pip is easier. 

Conda-forge is a community effort to build conda packages using free CI services. It is very convenient compared to building packages manually for all the platforms or compared to configuring CI services manually if you want to publish conda packages for more than one package. Also, the conda-forge channel is a one-stop-shop most 3rd party packages, which simplify management.

Honestly, I wish the CI services, especially AppVeyor, would be faster. But in my case, even 1-2 days for the packages to build is not a deal breaker.

In the lmfit case, the build was failing because emcee was not available for win64/python 3.6, the fact that lmfit has a test suite is irrelevant. There is nothing conda forge can do about it: the fact that all the dependencies you specify must be available to build your package is a conda feature, not conda-forge.
The fact that a pure python packages must be built for each platform/python version is another conda "feature". Maybe conda should have "source packages" for pure python packages,  it should have  suggested dependencies, or conda-build should try to recursively build packages for each dependency when a recipe is available (at least for pure-python packages). I don't know. You can suggest a solution, or simply not use conda if you don't find it convenient.

So, assuming one finds the conda ecosystem useful, the conda-forge if a big help. If you don't find conda useful, then yes conda-forge is pretty much useless too.

Going back to this PR, I opened a new PR #10 which will merge to the 0.9.5 branch. So I'm closing this.

**Comment by newville** [comment]
@tritemio,

On Thu, Mar 30, 2017 at 1:25 AM, Antonino Ingargiola <
notifications@github.com> wrote:

> @newville <https://github.com/newville>:
> A conda package for lmfit is useful if you (a) use conda/anaconda, (b)
> have lmfit as dependency of other conda packages. If lmfit is not a
> dependency then installing it by pip is easier.
>
> Conda-forge is a community effort to build conda packages using free CI
> services. It is very convenient compared to building packages manually for
> all the platforms or compared to configuring CI services manually if you
> want to publish conda packages for more than one package. Also, the
> conda-forge channel is a one-stop-shop most 3rd party packages, which
> simplify management.
>
Yes, I understand, appreciate, and use conda-forge.   I am a willing (and
voluntary) contributor.

> Honestly, I wish the CI services, especially AppVeyor, would be faster.
> But in my case, even 1-2 days for the packages to build is not a deal
> breaker.
>

Hm, well I think that if builds that took 2 minutes on a moderate local
machine took a full day on a cloud virtual service (even free), that is a
concern.


> In the lmfit case, the build was failing because emcee was not available
> for win64/python 3.6, the fact that lmfit has a test suite is irrelevant.
> There is nothing conda forge can do about it: the fact that all the
> dependencies you specify must be available to build your package is a conda
> feature, not conda-forge.
>

Uh, well the original recipe that failed on win-64 for Python 3.6 was
incorrect in that it said emcee is a dependency for running lmfit -- it is
not. So, there was something we could do about that.

But, that is not the only error that occurred.   There was also a conda
build error from Travis-CI (used for OSX) that was entirely due to a conda
build error, when downloading conda-forge dependencies -- it never even got
to downloading the lmfit package.  That can be blamed either on Travis  or
on conda-forge's build system.  The error looked like it came from
conda-forge.  It caused a false negative that was corrected by a trivial
re-try.  I have a hard time saying that conda-forge is flawless.

The fact that a pure python packages must be built for each platform/python
> version is another conda "feature".
>
Yeah.  The fact that it builds packages based on the success of a virtual
build is also a choice.   Not an unreasonable choice, but bound to have
false negatives.  Actually, the use of very 3 different virtual builds /
continuous integration tools is also a choice and a "feature".   Again, I
am not saying these choices are unreasonable, but neither of these is
actually inherent to the project goal of providing a single channel with
many contributed conda packages.

> Maybe conda should have "source packages" for pure python packages, it
> should have suggested dependencies, or conda-build should try to
> recursively build packages for each dependency when a recipe is available
> (at least for pure-python packages). I don't know. You can suggest a
> solution, or simply not use conda if you don't find it convenient.
>

It seems sort of strange to me that a complaint about a tool is met (by
both you and @ericdill) with "don't use it".  If I didn't want to use it, I
wouldn't complain, I would just walk away.  I know what it does (or claims
to do), what it does not claim to do, and why it is useful.   It is not
perfect. In fact, I have had more than one problem with false negatives
with it in the past two weeks.

I don't think conda-forge really tries to solve the problem of package
dependencies other than the existing requirements.txt, etc.   That's OK.

But, it does seem like a recipe for a pure python package could just be
`pip install package`.


> So, assuming one finds the conda ecosystem useful, the conda-forge if a
> big help. If you don't find conda useful, then yes conda-forge is pretty
> much useless too.
>

I want to use conda-forge.

> Going back to this PR, I opened a new PR #10
> <https://github.com/conda-forge/lmfit-feedstock/pull/10> which will merge
> to the 0.9.5 branch. So I'm closing this.
>

Why should there be another recipe for 0.9.5?   I thought there was already
one.

**Comment by tritemio** [comment]
@newville, the point of this PR was to build lmfit 0.9.5 for python 3.6 (but it failed due to a missing emcee package). The build was eventually done in #10.
